File: beer-game-spring/src/main/kotlin/com/beer/game/domain/Player.kt
package com.beer.game.domain

import com.beer.game.common.Role
import com.beer.game.events.DocumentType
import com.beer.game.events.Event
import com.beer.game.events.EventType
import com.beer.game.application.events.InternalEventListener
import java.util.UUID

data class Player(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val role: Role,
    var stock: Int,
    var backlog: Int,
    var weeklyOrder: Int,
    var lastOrder: Int,
    val cpu: Boolean,
    val orders: MutableList<Order>
) {
    companion object {
        fun retailPlayer(): Player {
            return Player(
                name = Role.RETAILER.toString(),
                role = Role.RETAILER,
                stock = 80,
                backlog = 80,
                weeklyOrder = 40,
                lastOrder = 40,
                cpu = false,
                orders = mutableListOf()
            )
        }

        fun wholesalerPlayer(): Player {
            return Player(
                name = Role.WHOLESALER.toString(),
                role = Role.WHOLESALER,
                stock = 1200,
                backlog = 1200,
                weeklyOrder = 600,
                lastOrder = 600,
                cpu = false,
                orders = mutableListOf()
            )
        }

        fun factoryPlayer(): Player {
            return Player(
                name = Role.FACTORY.name,
                role = Role.FACTORY,
                stock = 12000,
                backlog = 12000,
                weeklyOrder = 6000,
                lastOrder = 6000,
                cpu = false,
                orders = mutableListOf()
            )
        }
    }

    fun addOrder(order: Order) {
        orders.add(order)
    }

    fun hasEnoughStock(amount: Int): Boolean {
        return stock >= amount
    }

    fun emitCreation(listener: InternalEventListener, board: Board) {
        listener.publish(
            Event(
                document = board,
                documentId = board.id,
                entityId = id,
                documentType = DocumentType.PLAYER,
                eventType = EventType.NEW
            )
        )
    }

    fun emitUpdate(listener: InternalEventListener, boardId: String) {
        listener.publish(
            Event(
                document = this,
                documentId = boardId,
                entityId = id,
                documentType = DocumentType.PLAYER,
                eventType = EventType.UPDATE
            )
        )
    }
}

File: golang-beer-game/domain/Player.go
package domain

import (
	"errors"
	"fmt"
	"time"
)

type Player struct {
	Id          string
	Name        string
	Role        Role
	Stock       int64
	Backlog     int64
	WeeklyOrder int64
	LastOrder   int64
	Cpu         bool
	Orders      []Order
	CreatedAt   time.Time
}

type Role string

const (
	RoleRetailer   Role = "RETAILER"
	RoleWholesaler Role = "WHOLESALER"
	RoleFactory    Role = "FACTORY"
)

func CreateNewPlayer(role Role) Player {
	player := &Player{
		Name:   string(role),
		Role:   role,
		Cpu:    false,
		Orders: []Order{},
    CreatedAt: time.Now(),
	}
	switch role {
	case RoleRetailer:
		player.Stock = 80
		player.Backlog = 80
		player.WeeklyOrder = 40
		player.LastOrder = 40
	case RoleWholesaler:
		player.Stock = 1200
		player.Backlog = 1200
		player.WeeklyOrder = 600
		player.LastOrder = 600
	case RoleFactory:
		player.Stock = 12000
		player.Backlog = 1200
		player.WeeklyOrder = 6000
		player.LastOrder = 6000
	}
	return *player
}

func GetRole(role string) (Role, error) {
	switch role {
	case string(RoleRetailer):
		return RoleRetailer, nil
	case string(RoleWholesaler):
		return RoleWholesaler, nil
	case string(RoleFactory):
		return RoleFactory, nil
	}
	return RoleRetailer, errors.New(fmt.Sprintf("Given role is not supported %s", role))
}

func (p *Player) AddOrder(order Order) {
	p.Orders = append(p.Orders, order)
}

func (p *Player) HasStock(amount int64) bool {
	return p.Stock >= amount
}
